Implementation of Web client and Wire Mock

obedjamir opened the pull request:
Setup web client to consume an external api
get data from response and writes the data to a file
Implemented Wire Mock for testing

obedjamir committed:
Implementation of Web client and Wire Mock
Setup web client to consume an external api
get data from response and writes the data to a file
Implemented Wire Mock for testing

obedjamir committed:
Requested Changes Implemented
Class names changed based on proper naming convention
api call moved to kafka streams
relative changes made to MovieControllerTest

obedjamir committed:
Object Class Defined For MovieData
MovieData class used to store movie details
Try with resources imlemented in FileUtil
OmbdApiClient return Optional<>

PinakiGhosh commented on core/pom.xml:
```diff
@@ -0,0 +1,31 @@
+        <spring-kafka.version>2.8.5</spring-kafka.version>
+        <kafka-streams.version>3.1.0</kafka-streams.version>
+        <awaitility.version>4.2.0</awaitility.version>
+        <spring.cloud.version>2021.0.2</spring.cloud.version>
```
Dont add unnecessary properties

obedjamir replied:
this has been removed

PinakiGhosh commented on core/src/main/java/com/bankbazaar/webclient/core/model/MovieData.java:
```diff
@@ -0,0 +1,120 @@
+public class MovieData implements Serializable
+{
+    @JsonProperty(value = "Title")
+    @SerializedName(value = "Title")
```
all the names should be camel cased. **title** itself is fine. No need to convert it to **Title**

obedjamir replied:
I had to change the names because during deserialization @JsonProperty maps title to Title

PinakiGhosh commented on service/src/main/java/com/bankbazaar/webclient/service/config/KafkaFlow.java:
```diff
@@ -0,0 +1,112 @@
+                                    Optional<MovieData> response = omdbApiCLient.fetchMovieDetails(value);
+                                    if(response.isPresent()) {
+                                        if (fileUtil.createFile(response.get().getTitle()).exists()) {
+                                            response.get().setResponse(false);
```
I dont understand this

obedjamir replied:
This just checks if file exists

PinakiGhosh commented on service/src/main/java/com/bankbazaar/webclient/service/config/KafkaFlow.java:
```diff
@@ -0,0 +1,112 @@
+                                context ->
+                                {
+                                    log.error("retries exhausted",context.getLastThrowable());
+                                    MovieData response = omdbApiCLient.fetchMovieDetails(value).get();
```
This itself might throw out an error or the response may be null. In both cases you will get another exception

obedjamir replied:
Changes have been made to this code

PinakiGhosh commented on service/src/main/java/com/bankbazaar/webclient/service/producer/KafkaProducer.java:
```diff
@@ -0,0 +1,18 @@
+
+@Service
+public class KafkaProducer {
+    private  static  final  String TOPIC = "Create_File";
```
All this should be in application.properties

obedjamir replied:
Moved to properties files

PinakiGhosh commented on service/src/test/java/com/bankbazaar/webclient/service/controller/MovieControllerTest.java:
```diff
@@ -0,0 +1,57 @@
+    private String uri;
+    @Test
+    public void webClientTest() throws IOException {
+        WireMockServer wireMockServer = new WireMockServer();
```
You dont need to start it explicitly. Check the annotations which I have sent you

obedjamir replied:
@AutoConfigureWireMOck used

PinakiGhosh commented on service/src/test/java/com/bankbazaar/webclient/service/controller/MovieControllerTest.java:
```diff
@@ -0,0 +1,57 @@
+        HttpGet request = new HttpGet(uri+"/?t=iron-man&apikey=39e493d3");
+        HttpResponse httpResponse = httpClient.execute(request);
+
+        assertEquals(200, httpResponse.getStatusLine().getStatusCode());
```
Can you confirm if this test is working when your wifi connection is switched off?

obedjamir replied:
This dosen't work without internet connection

PinakiGhosh replied:
But that is the point of wiremock. That my test cases run without any dependency

obedjamir committed:
Requested Changes Implemented
Test cases implemented for KafkaFlow
Unnecessary properties removed
@AutoConfigureWireMock used

PinakiGhosh commented on service/src/main/java/com/bankbazaar/webclient/service/producer/KafkaProducer.java:
```diff
@@ -1,12 +1,14 @@
 public class KafkaProducer {
-    private  static  final  String TOPIC = "Create_File";
+    @Value("${kafka.producer.topic}")
+    private String TOPIC;
```
Variables should be in camel case

PinakiGhosh commented on service/src/main/resources/application.properties:
```diff
@@ -24,4 +24,5 @@ spring.cloud.stream.bindings.consumer-in-0.destination=File_Processor
 spring.cloud.stream.bindings.consumer-out-0.destination=Notification
 spring.cloud.stream.bindings.notification-in-0.destination=Notification
 omdb.api.uri=http://www.omdbapi.com
+kafka.producer.topic=Create_File
```
Property name should be meaning ful. Example what is this producing?

PinakiGhosh commented on service/src/test/java/com/bankbazaar/webclient/service/config/KafkaFlowTest.java:
```diff
@@ -0,0 +1,43 @@
+
+@SpringBootTest(properties = {"spring.kafka.bootstrap-servers=localhost:9092"}, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@ActiveProfiles({"test"})
+public class KafkaFlowTest {
```
Where is wiremock?

PinakiGhosh replied:
Also why are we pushing data directly to Kafka? Why not use the controller?

PinakiGhosh replied:
What about asserting the contains of the file?

obedjamir committed:
Updated Test Cases and Corrections
Test Cases only through controller using wire mock
File content validated
Response and Status class implemented
Changes made to Kafka Flow

obedjamir committed:
Proper Class Name And WireMock Implementation
Class names changed to proper names to signify function
Wire mock re-configured for testing
